credentials/sts/sts.go

```diff
@@ -128,17 +127,55 @@ func NewCredentials(opts Options) (credentials.PerRPCCredentials, error) {
 	return &callCreds{
 		opts:   opts,
 		client: makeHTTPDoer(roots),
-		bs:     makeBackoffStrategy(),
 	}, nil
 }
 
-// defaultBackoffStrategy returns an exponential backoff strategy based on the
-// default exponential backoff config, but with a smaller BaseDelay (because the
-// former is meant more for connection establishment).
-func defaultBackoffStrategy() backoff.Strategy {
-	config := grpcbackoff.DefaultConfig
-	config.BaseDelay = 100 * time.Millisecond
-	return backoff.Exponential{Config: config}
+// callCreds provides the implementation of call credentials based on an STS
+// token exchange.
+type callCreds struct {
+	opts   Options
+	client httpDoer
+
+	// Cached accessToken to avoid an STS token exchange for every call to
+	// GetRequestMetadata.
+	mu          sync.Mutex
+	cachedToken *tokenInfo
+}
+
+// GetRequestMetadata returns the cached accessToken, if available and valid, or
+// fetches a new one by performing an STS token exchange.
+func (c *callCreds) GetRequestMetadata(ctx context.Context, _ ...string) (map[string]string, error) {
+	if err := credentials.CheckSecurityLevel(ctx, credentials.PrivacyAndIntegrity); err != nil {
+		return nil, fmt.Errorf("unable to transfer STS PerRPCCredentials: %v", err)
+	}
+	if md := c.metadataFromCachedToken(); md != nil {
```

Comment:
Maybe just return the cached token (if valid) from this function, and share the logic to construct the metadata (below) so it only exists once?

```go
token := c.getCachedToken()
if token == nil {
  // get token; assign c.cachedToken
}
return map.....
```

Reply:
I made a helper function to return a map given a token (that way we only have that code once).

I didn't want to change the flow here since it currently keeps the more verbose code path at a minimal indentation, and indents the error handling, dealing with it first, which improves readability of the code.

Reply:
Done.